src/events/general.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Azure Storage Provider related event handlers."""
+
+
+import ops
+from charms.data_platform_libs.v0.azure_storage import AzureStorageProviderData
+from ops import CharmBase
+from ops.charm import ConfigChangedEvent, StartEvent
+
+from constants import AZURE_RELATION_NAME
+from core.context import Context
+from events.base import BaseEventHandler, compute_status
+from managers.azure_storage import AzureStorageManager
+from utils.logging import WithLogging
+
+
+class GeneralEvents(BaseEventHandler, WithLogging):
+    """Class implementing Azure Integration event hooks."""
+
+    def __init__(self, charm: CharmBase, context: Context):
+        super().__init__(charm, "general")
+
+        self.charm = charm
+        self.context = context
+
+        self.azure_provider_data = AzureStorageProviderData(self.charm.model, AZURE_RELATION_NAME)
+        self.azure_storage_manager = AzureStorageManager(self.azure_provider_data)
+
+        self.framework.observe(self.charm.on.start, self._on_start)
+        self.framework.observe(self.charm.on.update_status, self._on_update_status)
+
+        self.framework.observe(self.charm.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.charm.on.secret_changed, self._on_secret_changed)
+
+    @compute_status
+    def _on_start(self, _: StartEvent) -> None:
+        """Handle the charm startup event."""
+        pass
+
+    @compute_status
+    def _on_update_status(self, event: ops.UpdateStatusEvent):
+        """Handle the update status event."""
+        pass
+
+    @compute_status
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:  # noqa: C901
+        """Event handler for configuration changed events."""
+        # Only execute in the unit leader
+        if not self.charm.unit.is_leader():
+            return
+
+        self.logger.debug(f"Config changed... Current configuration: {self.charm.config}")
+        self.azure_storage_manager.update(self.context.azure_storage)
+
+    @compute_status
+    def _on_secret_changed(self, event: ops.SecretChangedEvent):
+        """Handle the secret changed event.
+
+        When a secret is changed, it is first checked that whether this particular secret
+        is used in the charm's config. If yes, the secret is to be updated in the relation
+        databag.
+        """
+        # Only execute in the unit leader
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.config.get("credentials"):
+            return
+
+        secret = event.secret
+        if self.charm.config.get("credentials") != secret.id:
+            return
```

Comment:
Same as for https://github.com/canonical/object-storage-integrators/pull/1/files#r1674101507

Reply:
On this regards, I think this is different from the secrets used for the charm relations. The user provides only secret ID when the charm is configured using `juju config`. As a result, the charm knows nothing but the secret ID. To get the user's label corresponding to the secret, the charm has to look up secrets using the secret ID.. but if unique lookup using the secret ID is possible, why not use that same lookup to check for the secret's existence at all? I agree that asking Juju team about this would be the best idea since they know it better. I will do that.

Reply:
I see your point and it is very valid. I'm thinking that you could (should) still stick a label on the secret internally (on the charm's level) perhaps, if the user didn't add one...? According to the old policy's mindset at least. 

Neverthless, as long as it works with ID, and there's no guidance from Juju team on this, for sure let's leave it as it is.

I'm very curious on the policy though, keep me in the loop for that pls.

Reply:
I have opened a thread in their Matrix server today [here](https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$nCMvXWPtF0dPHDMYd_xe6Ow0q1JXlkyF7UYbKjXjUJ4?via=ubuntu.com&via=matrix.org&via=matrix.debian.social). I'll keep you in the loop regarding this.